controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,371 @@
+					Group:   apiGroup.(string),
+					Version: version,
+					Kind:    kind.(string),
+				})
+			}
+		}
+	}
+
+	return &syncOnlys, finalErr
+}
+
+// Find the version that the constraint does not provided.
+// Constraint only provide kind and apiGroup. However the config resource need version
+func (r *ConstraintStatusReconciler) getAPIVersion(kind string, apiGroup string, skipRefresh bool) (string, error) {
+	// Cool time(10 min) to refresh discoveries
+	if len(r.apiResourceList) == 0 ||
+		r.discoveryLastRefreshed.Add(time.Minute*10).Before(time.Now()) {
+		err := r.refreshDiscoveryInfo()
+		if err != nil {
+			return "", err
+		}
+
+		// The discovery is just refeshed so skip another refesh
+		skipRefresh = true
+	}
+
+	for _, resc := range r.apiResourceList {
+		groupVerison, err := schema.ParseGroupVersion(resc.GroupVersion)
+		if err != nil {
+			r.Log.Error(err, "Cannot parse the group and version in getApiVersion ", "GroupVersion:", resc.GroupVersion)
+
+			continue
+		}
+
+		group := groupVerison.Group
+		version := groupVerison.Version
+		// Consider groupversion == v1 or groupversion == app1/v1
+		for _, apiResource := range resc.APIResources {
+			if apiResource.Kind == kind && group == apiGroup {
+				return version, nil
+			}
+		}
+	}
+
+	if !skipRefresh {
+		// Get new discoveryInfo, when any resource is not found
+		err := r.refreshDiscoveryInfo()
+		if err != nil {
+			return "", err
+		}
+
+		// Retry one more time after refresh the discovery
+		return r.getAPIVersion(kind, apiGroup, true)
+	}
+
+	return "", ErrNotFoundDiscovery
+}
+
+// Retrieve all groups and versions to add in config sync
+// Constraints present only kind and group so this function helps to find the version
+func (r *ConstraintStatusReconciler) refreshDiscoveryInfo() error {
+	r.discoveryLastRefreshed = time.Now()
+
+	discoveryClient := r.ClientSet.Discovery()
+
+	apiList, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		return err
+	}
+
+	// Save fetched discovery at apiResourceList
+	r.apiResourceList = apiList
+
+	return nil
+}
+
+func (r *ConstraintStatusReconciler) getUniqSyncOnly() []v1alpha1.SyncOnlyEntry {
+	syncOnlySet := map[v1alpha1.SyncOnlyEntry]bool{}
+	// Add to table for unique filtering
+	for _, syncEntries := range r.ConstraintToSyncOnly {
+		for _, entry := range syncEntries {
+			syncOnlySet[entry] = true
+		}
+	}
+
+	syncOnlys := make([]v1alpha1.SyncOnlyEntry, 0, len(syncOnlySet))
+	for key := range syncOnlySet {
+		syncOnlys = append(syncOnlys, key)
+	}
+
+	return syncOnlys
+}
+
+// This function is called when constraintPodStatus deleted event.
+// Delete constraintPodstatus key and refresh r.constraintMap with current constraint
+func (r *ConstraintStatusReconciler) handleDeleteEvent(
+	ctx context.Context, cpsName string, config *v1alpha1.Config,
+) error {
+	delete(r.ConstraintToSyncOnly, cpsName)
+
+	updatedSyncOnly := r.getUniqSyncOnly()
+
+	config.Spec.Sync.SyncOnly = updatedSyncOnly
```

Comment:
Are you able to make sure that the sync only entries actually changed before making the update API request?

For example, if there's two constraints that match pods, deleting just one will not cause the calculated sync only entries to change.

Reply:
We can do that but my concern is what if the versions are different?

ex: event v1 and beta